src/transformers/models/phi/modeling_phi.py

```diff
@@ -308,23 +310,25 @@ def forward(
         past_key_value: Optional[Cache] = None,
         output_attentions: bool = False,
         use_cache: bool = False,
+        **kwargs,
     ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
         bsz, q_len, _ = hidden_states.size()
 
-        # [batch_size, seq_length, 3 x hidden_size]
-        fused_qkv = self.query_key_value(hidden_states)
+        query_states = self.q_proj(hidden_states)
+        key_states = self.k_proj(hidden_states)
+        value_states = self.v_proj(hidden_states)
 
-        # 3 x [batch_size, seq_length, num_heads, head_dim]
-        (query_states, key_states, value_states) = self._split_heads(fused_qkv)
+        # Upcast queries and keys to prevent overflow in Phi-2
+        query_states = query_states.to(torch.float32)
+        key_states = key_states.to(torch.float32)
```

Comment:
```suggestion
        # Upcast queries and keys to prevent overflow in Phi-2 only for non-quantized models 
        if not hasattr(self.config, "_pre_quantization_dtype"):
            query_states = query_states.to(torch.float32)
            key_states = key_states.to(torch.float32)
```
if I am not mistaken the quantization primitives (linear8bit, linear4bit, GPTQLinear, AWQLinear), should automatically take care of casting the hidden states in half precision before the computation (but will add a small overhead)
What happens if you *not* upcast in fp32? does it always overflow? 
Maybe you can run a quick sanity check by loading the model in 4-bit by passing `load_in_4bit=True` in `from_pretrained` and confirm that it does not break the forward pass. Otherwise we can go for the fix I suggest here

Reply:
> What happens if you not upcast in fp32? does it always overflow?

It does not overflow when the model is in fp32, the error only occurs for fp16

Reply:
I went with the `attn_weights = torch.matmul(query_states.to(torch.float32), key_states.to(torch.float32).transpose(2, 3)) / math.sqrt(self.head_dim)` just to avoid storing them.

Reply:
Fine by me 🤗  we should recommend using `sdpa` but otherwise not an issue

Reply:
Thanks everyone! Just synced with main to fix the issue with the natten version. Tests should be passing in a while.